### PR TITLE
add basic pagination to remembers

### PR DESCRIPTION
**Pull request by olslash**
hello! this lets you do `?word 2` to retrieve page 2 of the remember for `word`. no more remembers lost to the character limit! namaste

### CODE REVIEW

**Comment by akalin** [comment]
:+1:

**Comment by rmmh** [comment]
Can you make the pagination occur at a comma boundary instead of a random byte?

**Comment by olslash** [comment]
yes, please hold

**Comment by olslash** [comment]
new code is up-- i will review it again tomorrow when my brain is more awake. But it seems to work! and should handle all the edge cases i think!

File: plugins/remember.py [hunk 1]
```diff
@@ -104,14 +105,68 @@ def forget(inp, chan='', db=None):
         return "I don't know about that."
 
 
-@hook.regex(r'^\? ?(.+)')
+def get_page(data, start_idx, min_page_len, max_page_len):
```
Comment (akalin) [inline plugins/remember.py]: maybe add a function comment, esp. explaining the return values

File: plugins/remember.py [hunk 2]
```diff
@@ -104,14 +105,68 @@ def forget(inp, chan='', db=None):
...
+    if last_comma_idx == -1 or last_comma_idx < min_page_len:
+        ret = data[start_idx:max_page_len + start_idx]
+        return ret, len(ret) + start_idx
+
+    end_idx = last_comma_idx + start_idx
+    return data[start_idx:end_idx], end_idx
+
+
+def get_pages(data, min_page_len, max_page_len):
+    result = {}
```
Comment (akalin) [inline plugins/remember.py]: my python is rusty, but shouldn't this be `[]`, since you want to return an array?

File: plugins/remember.py [hunk 3]
```diff
@@ -104,14 +105,68 @@ def forget(inp, chan='', db=None):
...
+    i = 0
+    last_idx = 0
+
+    while True:
+        page_data, last_idx = get_page(data, last_idx, min_page_len, max_page_len)
+
+        if page_data == "":
+            break
+
+        result[i] = page_data
```
Comment (akalin) [inline plugins/remember.py]: `result.append(page_data)` instead, to avoid the need for `i`?

File: plugins/remember.py [hunk 4]
```diff
@@ -104,14 +105,68 @@ def forget(inp, chan='', db=None):
...
+        i += 1
+
+    return result
+
+
+@hook.regex(r'^\? ?(\S+) ?(\d+)?')
 def question(inp, chan='', say=None, db=None):
-    "?<word> -- shows what data is associated with word"
+    "?<word> <page?>-- shows what data is associated with word"
```
Comment (akalin) [inline plugins/remember.py]: maybe mention optional page parameter

File: plugins/remember.py [hunk 5]
```diff
@@ -104,14 +105,68 @@ def forget(inp, chan='', db=None):
...
+
+    data = get_memory(db, chan, word.strip())
     if data:
-        say(data)
+        pages = get_pages(data, min_page_len, message_len_limit)
+        page_idx = page - 1 # 1-indexed
+        page_data = pages.get(page_idx)
+
+        if page_data is None:
+            return
```
Comment (akalin) [inline plugins/remember.py]: worth spitting out an error message like "foo has only X pages"?

File: plugins/remember.py [hunk 6]
```diff
@@ -104,14 +105,68 @@ def forget(inp, chan='', db=None):
...
+        page = 1
+
+    page = int(page)
+
+    data = get_memory(db, chan, word.strip())
     if data:
-        say(data)
+        pages = get_pages(data, min_page_len, message_len_limit)
+        page_idx = page - 1 # 1-indexed
+        page_data = pages.get(page_idx)
```
Comment (akalin) [inline plugins/remember.py]: ah yes, if you make `result` in `get_pages` an array, you'll have to change this

File: plugins/remember.py [hunk 7]
```diff
@@ -104,14 +105,68 @@ def forget(inp, chan='', db=None):
...
+
+        if page_data is None:
+            return
+
+        last_page = max(pages.keys())
```
Comment (akalin) [inline plugins/remember.py]: and this

File: plugins/remember.py [hunk 8]
```diff
@@ -104,14 +105,62 @@ def forget(inp, chan='', db=None):
...
+
+    data = get_memory(db, chan, word.strip())
     if data:
-        say(data)
+        pages = get_pages(data, min_page_len, message_len_limit)
+        page_idx = page - 1 # 1-indexed
+        try:
+            page_data = pages[page_idx]
+        except IndexError:
+            return
```
Comment (akalin) [inline plugins/remember.py]: still think you should say an error message or something here...

Comment (olslash) [inline plugins/remember.py]: I figured the precedent was not to, for example trying to retrieve a missing remember has no error response -- but we can do like "<word> only has 3 pages"?

@rmmh

Comment (akalin) [inline plugins/remember.py]: hmm good point. I guess I'm ok either way.

File: plugins/remember.py [hunk 9]
```diff
@@ -104,14 +105,71 @@ def forget(inp, chan='', db=None):
...
+    """
+    slices the data string, starting at start_idx, into a chunk no longer
+    than max_page_len. The slice will occur at a comma if the resulting
+    string would not be shorter than min_page_len.
+    returns a tuple of (result str, slice index)
+    """
+    page_data = data[start_idx:start_idx + max_page_len]
+    last_comma_idx = page_data.rfind(',')
+
+    if last_comma_idx == -1 or last_comma_idx < min_page_len:
```
Comment (rmmh) [inline plugins/remember.py]: don't need to check for -1, since -1 is < min_page_len

File: plugins/remember.py [hunk 10]
```diff
@@ -104,14 +105,71 @@ def forget(inp, chan='', db=None):
...
+    end_idx = last_comma_idx + start_idx
+    return data[start_idx:end_idx], end_idx
+
+
+def get_pages(data, min_page_len, max_page_len):
+    "slices the data string into pages using get_page, returning a list"
+    result = []
+
+    page_data, last_idx = get_page(data, 0, min_page_len, max_page_len)
+    while page_data != '':
```
Comment (rmmh) [inline plugins/remember.py]: `while page_data`

File: plugins/remember.py [hunk 11]
```diff
@@ -104,14 +105,71 @@ def forget(inp, chan='', db=None):
...
+
+    return result
+
+more_pages_message = ' (%s page(s) left)'
+page_missing_message = "%s only has %s page(s)"
+
+@hook.regex(r'^\? ?(\S+) ?(\d+)?')
 def question(inp, chan='', say=None, db=None):
-    "?<word> -- shows what data is associated with word"
+    "?<word> <page?>-- shows what data is associated with word, paginated by page (1 if omitted)"
```
Comment (rmmh) [inline plugins/remember.py]: `?<word> [page] -- shows data associated with word, paginated by page (default 1)`

File: plugins/remember.py [hunk 12]
```diff
@@ -104,14 +105,71 @@ def forget(inp, chan='', db=None):
...
     db_init(db)
 
-    data = get_memory(db, chan, inp.group(1).strip())
+    message_len_limit = 512 - len(more_pages_message) - 75 # fudge factor for protocol overhead
+    min_page_len = 100
+
+    word = inp.group(1).strip()
+    page = inp.group(2)
```
Comment (rmmh) [inline plugins/remember.py]: `page = inp.group(2) or 1`

File: plugins/remember.py [hunk 13]
```diff
@@ -187,3 +245,68 @@ def test_forget(self):
...
+        assert get_page('123456789', 0, 5, 8) == ('12345678', 8) # max length
+        assert get_page('123,456789', 0, 5, 8) == ('123,4567', 8) # min length not satisfied
+        assert get_page('123,45,67,89', 0, 5, 8) == ('123,45', 6) # chooses correct comma
+        assert get_page('', 0, 5, 8) == ('', 0) # correct empty result
+        # start offsets
+        assert get_page('123456789', 1, 5, 9) == ('23456789', 9)
+        assert get_page('123456789', 1, 5, 8) == ('23456789', 9)
+        assert get_page('1234567,89', 1, 5, 8) == ('234567', 7)
+        assert get_page('', 1, 5, 8) == ('', 1)
+
```
Comment (rmmh) [inline plugins/remember.py]: extra newline

File: plugins/remember.py [hunk 14]
```diff
@@ -187,3 +245,68 @@ def test_forget(self):
...
+        for _ in range(0, 1000):
+            long_string += 'a'
+
+        self.remember(long_string, chan='#long')
+        assert more_pages_message % 2 in self.question('long', chan='#long') # "2 pages left"
+        assert more_pages_message % 1 in self.question('long 2', chan='#long')
+        assert not re.match(r'\d', self.question('long 3', chan='#long'))
+
+    def test_paging_result(self):
+        long_string = 'long '
```
Comment (rmmh) [inline plugins/remember.py]: `long_string = 'long ' + 'x' * 300 + 'y' * 300 + 'z' * 300`

File: plugins/remember.py [hunk 15]
```diff
@@ -187,3 +245,68 @@ def test_forget(self):
...
+        for _ in range(0, 300):
+            long_string += 'y'
+        for _ in range(0, 300):
+            long_string += 'z'
+
+        self.remember(long_string, chan='#long')
+        p1_data = self.question('long', chan='#long')
+        p2_data = self.question('long 2', chan='#long')
+        p3_data = self.question('long 3', chan='#long')
+
```
Comment (rmmh) [inline plugins/remember.py]: all_data = p1_data + p2_data + p3_data
    assert all_data.count('x') == 300
    ...

File: plugins/remember.py [hunk 16]
```diff
@@ -187,3 +245,68 @@ def test_forget(self):
...
+        count_y = p1_data.count('y') + p2_data.count('y') + p3_data.count('y')
+        count_z = p1_data.count('z') + p2_data.count('z') + p3_data.count('z')
+
+        # no data lost
+        assert count_x == 300
+        assert count_y == 300
+        assert count_z == 300
+
+    def test_missing_page(self):
+        long_string = ''
```
Comment (rmmh) [inline plugins/remember.py]: `long_string = 'x' * 1000`

File: plugins/remember.py [hunk 17]
```diff
@@ -187,3 +245,68 @@ def test_forget(self):
...
+
+
+    def test_get_pages(self):
+        assert get_pages('123456789', 5, 8) == ['12345678', '9']
+        assert get_pages('123,456789', 5, 8) == ['123,4567', '89']
+        assert get_pages('123,45,67,89', 5, 8) == ['123,45', ',67,89']
+        assert get_pages('', 5, 8) == []
+
+    def test_paging_message(self):
+        long_string = 'long '
```
Comment (rmmh) [inline plugins/remember.py]: `long_string = 'long ' + 'a' * 1000`

**Comment by olslash** [comment]
thanks for the help @rmmh -- I made your changes :)

File: plugins/remember.py [hunk 18]
```diff
@@ -104,14 +105,69 @@ def forget(inp, chan='', db=None):
...
+    return result
+
+more_pages_message = ' (%s page(s) left)'
+page_missing_message = "%s only has %s page(s)"
+message_len_limit = 512 - len(more_pages_message) - 75 # fudge factor for protocol overhead
+
+@hook.regex(r'^\? ?(\S+) ?(\d+)?')
 def question(inp, chan='', say=None, db=None):
-    "?<word> -- shows what data is associated with word"
+    "?<word> page] -- shows what data is associated with word, paginated by page (default 1)"
```
Comment (rmmh) [inline plugins/remember.py]: [page]

**Pull request merged**